omnibor/src/input_manifest.rs

```diff
@@ -0,0 +1,129 @@
+//! [`InputManifest`] type that represents build inputs for an artifact.
+
+use crate::{ArtifactId, SupportedHash};
+use gitoid::{Blob, HashAlgorithm, ObjectType};
+use std::fmt::{Display, Formatter, Result as FmtResult};
+
+/// A manifest describing the inputs used to build an artifact.
+///
+/// The manifest is constructed with a specific target artifact in mind.
+/// The rest of the manifest then describes relations to other artifacts.
+/// Each relation can be thought of as describing edges between nodes
+/// in an Artifact Dependency Graph.
+///
+/// Each relation encodes a kind which describes how the other artifact
+/// relates to the target artifact.
+///
+/// Relations may additionally refer to the [`InputManifest`] of the
+/// related artifact.
+#[derive(Debug)]
+pub struct InputManifest<H: SupportedHash> {
+    /// The artifact the manifest is describing.
+    target: ArtifactId<H>,
+
+    /// The relations recorded in the manifest.
+    relations: Vec<Relation<H>>,
+}
+
+impl<H: SupportedHash> InputManifest<H> {
+    /// Get the ID of the artifact this manifest is describing.
+    #[inline]
+    pub fn target(&self) -> ArtifactId<H> {
+        self.target
+    }
+
+    /// Get the relations inside an [`InputManifest`].
+    #[inline]
+    pub fn relations(&self) -> &[Relation<H>] {
+        &self.relations[..]
+    }
+}
+
+impl<H: SupportedHash> Display for InputManifest<H> {
+    // The OmniBOR spec actually specifies using _only_ `\n` for newlines,
+    // regardless of the host platform specification, so Clippy's recommendation
+    // here would cause our code to violate the spec.
+    #[allow(clippy::write_with_newline)]
+    #[inline]
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        // Per the spec, this prefix is present to substantially shorten
+        // the metadata info that would otherwise be attached to all IDs in
+        // a manifest if they were written in full form. Instead, only the
+        // hex-encoded hashes are recorded elsewhere, because all the metadata
+        // is identical in a manifest and only recorded once at the beginning.
+        write!(f, "gitoid:{}:{}\n", Blob::NAME, H::HashAlgorithm::NAME)?;
+
```

Comment:
Should we write the `target` here so that the user can see that info as well? I am going to be honest, this is my first actual code review in Rust, so sorry if I have bad questions/suggestions.

Reply:
Yeah this is definitely a good question. Right now (and I believe this is consistent with the spec), the target ID isn't persisted in the manifest. I believe the idea is that the artifact itself embeds the manifest ID and perhaps the manifest title itself _is_ some encoding of the target ID.

That said, I think it probably _should_ encode the target ID. This way the manifest/target linkage becomes "doubly-linked" (manifest refers to target ID, target refers to manifest ID).

The one constraint is that this creates a chicken-and-egg problem. The manifest has its own ID based on its contents. If those contents include the target, and the target includes the manifest ID, then how do you break the chain? One of them has to be identifiable while _ignoring_ the ID of the other. It's probably right and proper that the manifest Artifact ID ignore the reference to the target ID.

Reply:
Thinking more, maybe this is where the suggestion of making an encoding of the target ID be the title of the manifest file (which inherently isn't part of the hash of its contents used to produce the manifest's Artifact ID) came from. It solves the chicken and egg problem _without_ resorting to special-casing construction of the Artifact ID to be "usually we hash the whole file, except when the file is an Artifact Input Manifest, in case we hash the whole file _except_ any reference to the target's Artifact ID.

So the workflow becomes, for a specific target artifact.

1. During artifact production, produce Artifact ID's for every input
2. Construct the Artifact Input Manifest
3. Derive the Artifact ID of the Artifact Input Manifest
4. Embed a reference to the manifest's Artifact ID into the target artifact
5. Derive the Artifact ID of the target artifact
6. Update the title of the Artifact Input Manifest file to be a file-system-safe encoding of the target artifact's Artifact ID

Reply:
(This also suggests to me that `ArtifactId` is missing some mechanism to get a file-system-safe encoding of itself. Something which only uses characters file systems would be fine with in file names with minimal issue. Probably just means replacing `:` with `_` or `-`)

Reply:
Further thinking (my apologies for multiple messages here haha) is that the API exposed by the `omnibor` crate should likely be something like:

```rust
let my_artifact = File::open("blah.rs")?;
ManifestBuilder::for_artifact(&my_artifact)
    .add_input(/* some input artifact */)
    .add_input(/* some input artifact */)
    .finish()?;
```

Where `finish` basically does the steps I outlined above correctly, in terms of the order of operations for creating the manifest, updating the artifact, then updating the manifest. So everything refers to each other in the correct way.

Reply:
Okay I've been ruminating for a little bit, and I feel more confident that the means of breaking the chain I described above is the right one. It retains the key attribute that the Artifact ID is a cryptographic commitment to _both_ the contents of the artifact at a specific time and the state of the known artifact dependency graph formed by the implicit Merkel Tree of the manifest and the manifests of the input artifacts.

Reply:
Yep, that makes sense to me. If we want to know that a specific set of inputs created a specific target and allow that target to be embedded with a hash of the list of inputs and the target that we embedding, we end up with a chicken-and-egg problem. For some reason I had actually thought it was in the spec that we are supposed to have the target in the input manifest, but looking again I misremembered.

Reply:
Just a question and you can totally say no, but should we add it to print out all fields for this struct, or is the intent to only display the things that will end up in the input manifest?

Reply:
1. Regarding printing all elements, the `Debug` impl (which is auto-derived), does that; so printing like the following will show everything:

```rust
println!("{:?}", my_manifest);
```

You can also do pretty-printing with the `:#?` format string instead of `:?`.

2. I think the proposed solution of making the filename contain both the target Artifact ID _and_ the manifest Artifact ID makes sense. The one thing we'd probably want to plan around is file name / path length limits. I know this comes up for systems on Windows, for example. We'll really need to plan for it anyway, it just becomes more pressing with longer names.